File: array-hashing/hashmap.py
from __future__ import annotations
from typing import Optional, List


class Node:
    def __init__(self, key: int = -1, val: int = -1, next: Optional[Node] = None) -> None:
        self.key = key
        self.val = val
        self.next = next


class MyHashMap:
    def __init__(self) -> None:
        self.map: List[Node] = []
        for _ in range(1000):
            self.map.append(Node())

    def put(self, key: int, val: int)-> None:
        cur = self.map[self.key(key)]

        while cur.next:
            if cur.next.key == key:
                cur.next.val = val
                return
            cur = cur.next

        cur.next = Node(key, val)

    def retrieve(self, key: int) -> int:
        cur: Node = self.map[self.key(key)]
        while cur.next:
            if cur.next.key == key:
                return cur.val
            cur = cur.next
        return -1
      
    def remove(self, key: int) -> None:
      cur: Node = self.map[self.key(key)]
      while cur.next:
        if cur.next.key == key:
          cur.next = cur.next.next
          return
        cur = cur.next

    def key(self, key: int) -> int:
        return key % len(self.map)

File: array-hashing/hashmap.go
package main

type Node struct {
	key  int
	val  int
	next *Node
}

type MyHashMap struct {
	buckets []*Node
	size    int
}

func Constructor() *MyHashMap {
	return &MyHashMap{
		buckets: make([]*Node, 1000),
		size:    1000,
	}
}

// put
func (m *MyHashMap) Put(key int, val int) {
	bucket := m.key(key)
	head := m.buckets[bucket]
	if head == nil {
		head = &Node{
			key:  key,
			val:  val,
			next: nil,
		}
		return
	}
	cur := head
	for {
		if cur.key == key {
			cur.val = val
			return
		}
		if cur.next == nil {
			cur.next = &Node{key: key, val: val}
			return
		}
		cur = cur.next
	}
}

// get
func (m *MyHashMap) Get(key int) int {
	bucket := m.key(key)
	cur := m.buckets[bucket]

	for cur != nil {
		if cur.key == key {
			return cur.val
		}
		cur = cur.next
	}
	return -1
}

// remove
func (m *MyHashMap) Remove(key int) {
	idx := m.key(key)
	cur := m.buckets[idx]
	if cur == nil {
		return
	}
	// If head is the key, remove head
	if cur.key == key {
		m.buckets[idx] = cur.next
		return
	}
	prev := cur
	cur = cur.next
	for cur != nil {
		if cur.key == key {
			prev.next = cur.next
			return
		}
		prev, cur = cur, cur.next
	}
}

// key
func (m *MyHashMap) key(key int) int {
	return key % len(m.buckets)
}
